### ISSUE TITLE
Sampling from a truncated MixtureModel

### ISSUE THREAD

**Issue by mirkobunse** [JuliaStats/Distributions.jl#1039]
I can not sample from a truncated MixtureModel. Calling
```julia
rand(truncated(MixtureModel([Normal(), Normal()]), -1, 1), 3)
```
produces the following stack trace in version `v0.21.11`:

    ERROR: MethodError: no method matching iterate(::MixtureModel{Univariate,Continuous,Normal{Float64},Float64})
    Closest candidates are:
      iterate(::Core.SimpleVector) at essentials.jl:589
      iterate(::Core.SimpleVector, ::Any) at essentials.jl:589
      iterate(::ExponentialBackOff) at error.jl:171
      ...
    Stacktrace:
     [1] copyto!(::Array{Float64,1}, ::MixtureModel{Univariate,Continuous,Normal{Float64},Float64}) at ./abstractarray.jl:646
     [2] _collect(::UnitRange{Int64}, ::MixtureModel{Univariate,Continuous,Normal{Float64},Float64}, ::Base.HasEltype, ::Base.HasLength) at ./array.jl:563
     [3] collect(::MixtureModel{Univariate,Continuous,Normal{Float64},Float64}) at ./array.jl:557
     [4] #quantile#51(::Bool, ::Function, ::MixtureModel{Univariate,Continuous,Normal{Float64},Float64}, ::Float64) at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.0/Statistics/src/Statistics.jl:911
     [5] quantile(::MixtureModel{Univariate,Continuous,Normal{Float64},Float64}, ::Float64) at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.0/Statistics/src/Statistics.jl:911
     [6] quantile(::Truncated{MixtureModel{Univariate,Continuous,Normal{Float64},Float64},Continuous,Float64}, ::Float64) at /home/bunse/.julia/packages/Distributions/0vgDZ/src/truncate.jl:90
     [7] rand(::Random.MersenneTwister, ::Truncated{MixtureModel{Univariate,Continuous,Normal{Float64},Float64},Continuous,Float64}) at /home/bunse/.julia/packages/Distributions/0vgDZ/src/univariates.jl:175
     [8] rand! at /home/bunse/.julia/packages/Distributions/0vgDZ/src/univariates.jl:165 [inlined]
     [9] rand at /home/bunse/.julia/packages/Distributions/0vgDZ/src/univariates.jl:158 [inlined]
     [10] rand(::Truncated{MixtureModel{Univariate,Continuous,Normal{Float64},Float64},Continuous,Float64}, ::Int64) at /home/bunse/.julia/packages/Distributions/0vgDZ/src/genericrand.jl:24
     [11] top-level scope at none:0

Feasible, but less convenient, is to sample from a MixtureModel of truncated Distributions. As far as I understand, this second approach should be equivalent to the above one:

```julia
rand(MixtureModel([truncated(Normal(), -1, 1), truncated(Normal(), -1, 1)]), 3)
```

    3-element Array{Float64,1}:
      0.07164009677046827 
      0.7757187289191435  
     -0.9822884277105955  

Maybe the function `truncated(::MixtureModel, l, u)` (first approach) can be changed to produce a MixtureModel with truncated components, i.e. the result of the second approach, automagically?

I hope that, at least, the workaround just mentioned can help other users with the same problem.

**Comment by matbesancon** [comment]
> Maybe the function truncated(::MixtureModel, l, u) (first approach) can be changed to produce a MixtureModel with truncated components, i.e. the result of the second approach, automagically?

This seems fine yes, PR welcome, even though I think some other solution might exist, some method that we didn't implement for MixtureModel

**Comment by matbesancon** [comment]
Ok yes I see there seems to be no `rand` implemented for `MixtureModel`, which is odd.
I think
```julia
rand(rng::AbstractRNG, d::MixtureModel) =
    rand(rng, sampler(d))
```

would fix it

**Comment by nignatiadis** [comment]
https://github.com/JuliaStats/Distributions.jl/pull/1195 fixes this by implementing `quantile` for `MixtureModel`. On that branch:
```julia
julia> rand(truncated(MixtureModel([Normal(), Normal()]), -1, 1), 3)
3-element Array{Float64,1}:
 -0.537416790913867
 -0.7758280778450812
  0.5279238107316366
```

**Comment by devmotion** [comment]
Fixed in Distributions 0.25.16.
